docs/docs/explanations/00_noir/13_traits.md

```diff
@@ -0,0 +1,348 @@
+
+```self
+trait Second {
+    fn second(self) -> Field;
+}
+
+impl<T> Second for (T, Field) {
+    fn second(self) -> Field {
+        self.1
+    }
+}
+```
+
+You can also implement a trait for every type this way:
+
+```rust
+trait Debug {
+    fn debug(self);
+}
+
+impl<T> Debug for T {
+    fn debug(self) {
+        println(self);
+    }
+}
+
+fn main() {
+    1.debug();
+}
+```
+
+## Impl Specialization
+
+When implementing traits for a generic type it is possible to implement the trait for only a certain combination
+of generics. This can be either as an optimization or because those specific generics are required to implement the trait.
+
+```rust
+trait Sub {
+    fn sub(self, other: Self) -> Self;
+}
+
+struct NonZero<T> {
+    value: T,
+}
+
+impl Sub for NonZero<Field> {
+    fn sub(self, other: Self) -> Self {
+        let value = self.value - other.value;
+        assert(value != 0);
+        NonZero { value }
+    }
+}
+```
+
+## Overlapping Implementations
+
+Overlapping implementations are disallowed by Noir to ensure Noir's decision on which impl to select is never ambiguous.
+This means if a trait `Foo` is already implemented
+by a type `Bar<T>` for all `T`, then we cannot also have a separate impl for `Bar<Field>` (or any other
+type argument). Similarly, if there is an impl for all `T` such as `impl<T> Debug for T`, we cannot create
+any more impls to `Debug` for other types since it would be ambiguous which impl to choose for any given
+method call.
+
+```rust
+trait Trait {}
+
+// Previous impl defined here
+impl<A, B> Trait for (A, B) {}
+
+// error: Impl for type `(Field, Field)` overlaps with existing impl
+impl Trait for (Field, Field) {}
+```
+
+## Trait Coherence
+
+Another restriction on trait implementations is coherence. This restriction ensures other crates cannot create
+impls that may overlap with other impls, even if several unrelated crates are used as dependencies in the same
+program.
+
+The coherence restriction is: to implement a trait, either the trait itself or the object type must be declared
+in the crate the impl is in.
+
+In practice this often comes up when using types provided by libraries. If a library provides a type `Foo` that does
+not implement a trait in the standard library such as `Default`, you may not `impl Default for Foo` in your own crate.
+While restrictive, this prevents later issues or silent changes in the program if the `Foo` library later added its
+own impl for `Default`. If you are a user of the `Foo` library in this scenario and need a trait not implemented by the
+library your choices are to either submit a patch to the library or use the newtype pattern.
+
+### The Newtype Pattern
+
+The newtype pattern gets around the coherence restriction by creating a new wrapper type around the library type
+that we cannot create `impl`s for. Since the new wrapper type is defined in our current crate, we can create
+impls for any trait we need on it.
```

Comment:
Does this read alright? I'm thinking of removing the "that we cannot create `impl`s for" part since it can be interpreted as "the newtype we cannot create impls for" even though it is specified in the later paragraph.

Reply:
I think that it should be clear enough in context of the coherence restriction. You can only make this mistake if you don't know the problem which it is trying to work around.